text/0000-if-let-chains.md

```diff
@@ -0,0 +1,2000 @@
+        }
+    }
+
+    if let PAT_3 = EXPR_3 {
+        if EXPR_4 {
+            break 'FRESH_LABEL { EXPR_ELSE_IF }
+        }
+    }
+
+    { EXPR_ELSE }
+}
+```
+
+Having an `else` branch is optional.
+The following example without an `else` branch:
+
+```rust
+if let PAT_1 = EXPR_1 &&
+   let PAT_2 = EXPR_2 {
+    EXPR_IF
+}
+```
+
+is simply desugared into:
+
+```rust
+if let PAT_1 = EXPR_1 {
+    if let PAT_2 = EXPR_2 {
+        EXPR_IF
+    }
+}
+```
+
+If we have an `else if` branch but no `else` branch, such as in this example:
+
+```rust
+if let PAT_1 = EXPR_1 &&
+   let PAT_2 = EXPR_2 {
+    EXPR_IF
+} else if let PAT_3 = EXPR_3 &&
+       EXPR_4 {
+    EXPR_ELSE_IF
+}
+```
+
+the semantics are defined by the following desugaring:
+
+```rust
+'FRESH_LABEL: {
+    if let PAT_1 = EXPR_1 {
+        if let PAT_2 = EXPR_2 {
+            break 'FRESH_LABEL { EXPR_IF }
+        }
+    }
+
+    if let PAT_3 = EXPR_3 {
+        if EXPR_4 {
+            break 'FRESH_LABEL { EXPR_ELSE_IF }
+        }
+    }
+}
+```
+
+## `while let`-chains
+
+The semantics of `while let`-chains can be understood by an in-surface-language
+desugaring using only [RFC 2046], `loop` and `if let`.
+
+For example:
+
+```rust
+while EXPR_1 &&
+      let PAT_2 = EXPR_2 &&
+      let PAT_3 = EXPR_3 &&
+      EXPR_4 {
+    EXPR_WHILE
+}
+```
+
+is defined by desugaring into:
+
+```rust
+loop {
+    if EXPR_1 &&
+       let PAT_2 = EXPR_2 &&
+       let PAT_3 = EXPR_3 &&
+       EXPR_4 {
+        { EXPR_WHILE }
+        continue;
+    }
+    break;
+}
+```
+
+This desugaring relies on the previously discussed desugaring for `if let`-chains.
+
+More generally, we may desugar:
+
+```rust
+while in_if_list {
+    EXPR_WHILE
+}
+```
+
+into:
+
+```rust
+loop {
+    if in_if_list {
+        { EXPR_WHILE }
+        continue;
+    }
+    break;
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This RFC mandates additions to the grammar as well as adding syntax lowering
+passes. These are small additions, but nonetheless the language specification
+is possibly made more complex by it. While this complexity will be used by some
+and therefore, the RFC argues, motivates the added complexity, it will not be
+used all users of the language. However, as discussed in the [motivation],
+by unifying constructs in the language conceptually and grammatically,
+we may also say that complexity is *reduced*.
+
+When it comes to `if let`-chains, the feature is already supported by the
+macro `if_chain!`. Some may feel that this is enough.
+
+It should also be taken into account that some breakage will occur as a result
+of this RFC. Sergio Benitez has however done some review of the crates.io
+ecosystem and found zero cases of actual breakage. At any rate, writing
+`let PAT = EXPR && ..` as a user is a bad thing to do.
+
+Finally, some may argue that this is *"a lot of ad-hoc syntax to deprecate when
+the proper solution solving all the listed problems is implemented"*.
```

Comment:
What are the phrases "ad-hoc syntax" and "the proper solution" trying to refer to in this sentence?

Everything else in the RFC makes perfect sense and seems compelling to me.